src/storage/backend/sqlite/SQLite.cc

```diff
@@ -279,36 +314,109 @@ OperationResult SQLite::DoErase(ResultCallback* cb, ValPtr key) {
     if ( ! key_data )
         return {ReturnCode::SERIALIZATION_FAILED, "Failed to serialize key"};
 
-    auto stmt = erase_stmt.get();
+    auto stmt = unique_stmt_ptr(erase_stmt.get(), sqlite3_reset);
 
-    if ( auto res = CheckError(sqlite3_bind_blob(stmt, 1, key_data->data(), key_data->size(), SQLITE_STATIC));
+    if ( auto res = CheckError(sqlite3_bind_blob(stmt.get(), 1, key_data->data(), key_data->size(), SQLITE_STATIC));
          res.code != ReturnCode::SUCCESS ) {
-        sqlite3_reset(stmt);
         return res;
     }
 
-    return Step(stmt, false);
+    return Step(stmt.get(), false);
 }
 
 /**
  * Removes any entries in the backend that have expired. Can be overridden by
  * derived classes.
  */
 void SQLite::DoExpire(double current_network_time) {
-    auto stmt = expire_stmt.get();
+    int status;
+    char* errMsg = nullptr;
+    unique_stmt_ptr stmt;
+
+    // Begin an exclusive transaction here to lock the database for this one process. That
+    // will ensure there isn't a TOCTOU bug in the time check below.
+    status = SQLITE_BUSY;
+    while ( status == SQLITE_BUSY ) {
+        status = sqlite3_exec(db, "begin immediate transaction", nullptr, nullptr, &errMsg);
+        sqlite3_free(errMsg);
+
+        if ( status == SQLITE_OK )
+            break;
+        else
+            // If any other status is returned here, give up. Notably, this includes
+            // SQLITE_BUSY which will be returned if there was already a transaction
+            // running. If one node got in and made the transaction, expiration is
+            // happening so the rest don't need to retry.
+            return;
+    }
 
-    int status = sqlite3_bind_double(stmt, 1, current_network_time);
+    // Automatically rollback the transaction when this object is deleted.
+    auto deferred_rollback = util::Deferred([this]() {
+        char* errMsg = nullptr;
+        sqlite3_exec(db, "rollback transaction", nullptr, nullptr, &errMsg);
+        sqlite3_free(errMsg);
+    });
+
+    // Check if the expiration control key is less than the interval. Exit if not.
+    stmt = unique_stmt_ptr(get_expiry_last_run_stmt.get(), sqlite3_reset);
+    while ( status != SQLITE_ROW ) {
+        status = sqlite3_step(stmt.get());
+        if ( status == SQLITE_ROW ) {
+            double last_run = sqlite3_column_double(stmt.get(), 0);
+
+            DBG_LOG(DBG_STORAGE, "Expiration last run: %f  diff: %f  interval: %f", last_run,
+                    current_network_time - last_run, zeek::BifConst::Storage::expire_interval);
+
+            if ( current_network_time > 0 &&
+                 (current_network_time - last_run) < zeek::BifConst::Storage::expire_interval )
+                return;
+        }
+        else
+            return;
+    }
+
+    // Update the expiration control key
```

Comment:
Still not sure if we should really do this regularly. My comment about a disk spinning up was very tongue-in-cheek, but I do think it is worth considering completely idle systems (QA appliances, non-connected production appliances or standby-systems, virtual machines, maybe idle Windows systems with Zeek embedded, ...). It is quite unfortunate that having an SQLite backend open would mean a database on disk is written to every 15 seconds.

Not stubborn, but it is important enough to me to mention it one more time :-) Feel free to resolve, obviously.